content/docs/guides/ha_scale/perf-scale.md

```diff
@@ -0,0 +1,121 @@
+---
+title: "Performance and Scalability"
+description: "Performance benchmark on control plane and data plane sidecar and scalability test of Open Service Mesh"
+type: docs
+---
+
+Open Service Mesh aims to create a high-performance service mesh while providing all the essential service mesh features. This document describes the tests performed to evaluate the performance and scalability of OSM.
+
+## Testing environment
+
+This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory). It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.
+
+## Process
+
+The testing mesh contains two types of service. One is a load generator (modified from [wrk2](https://github.com/giltene/wrk2)), which can send requests at specified rate, load balanced among all other services in the mesh. The load generator serves as the client side of a request. Throughout our test, there will be at most one load generator service. The other type of service is a simple echo application that replies with whatever it receives from an inbound request. It serves as the server side of a request. We can deploy multiple copies of the echo application in the mesh. In short, the mesh topology is a typical 1-to-N service architecture.
+
+<p align="center">
+  <img src="/docs/images/perf/mesh-topo.png" width="650"/>
+</p>
+
+
+## Performance
+
+In this test, we focus on the latency overhead and resource consumption added by OSM's. We currently test the per pod RPS at 100, 500, 1000, 2000 and 4000 respectively.
+
+### Latency
+
+Below is the request latency at different RPS levels with and without OSM installed. Note, with OSM installed the added latency comes from both the client and server sidecar.
+
+<p align="center">
+  <img src="/docs/images/perf/wo-osm-latency.png" width="650"/>
+</p>
+
+<p align="center">
+  <img src="/docs/images/perf/w-osm-latency.png" width="650"/>
+</p>
+
+<table style="width: 100%">
+<thead>
+<tr>
+<th>Per pod RPS</th>
+<th>100</th>
+<th>500</th>
+<th>1000</th>
+<th>1500</th>
+<th>2000</th>
+<th>4000</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>P99 latency without OSM</td>
+<td>2</td>
+<td>4.6</td>
+<td>3.7</td>
+<td>4.23</td>
+<td>4.3</td>
+<td>3.1</td>
+</tr>
+<tr>
+<td>P99 latency with OSM</td>
+<td>10.1</td>
+<td>4.93</td>
+<td>5.93</td>
+<td>5</td>
+<td>17.5</td>
+<td>32.8</td>
```

Comment:
You sure these higher numbers at 4000 rps, still aren't due to cpu bottlenecks?

Reply:
Yes. We tested the 8000 rps and the latency continues to increase.

Reply:
The sidecar CPU usage is below in the Resource Consumption section that can be compared with the latency.

Reply:
what limits did you use? I'd also be careful that with the p99.9 latency you may have a spike in CPU usage for brief moment that hits limits, and affects low number of requests to put this over the limit. so we may want to view cpu usage over time for each container

Reply:
The CPU limit for echo application is 1 core. P99.9 is averaged across the collected metrics in the 10 mins testing time. As our metrics pulling interval is 15s, we will have 60 data points. I don't think it's necessary to show the over-time CPU usage chart in this document, as it's too detailed.

Reply:
For reference, this is the sidecar CPU usage over a test run. It looks very stable.

![image](https://user-images.githubusercontent.com/872876/160935357-010cdfdb-0158-43bb-8016-6a280099d9a4.png)

Reply:
can you chart plot the max value? that might be more indicative of anything hitting limits.

what is the container limit btw?

Reply:
Discussed offline and we noticed that the increase in latency is caused due to the Envoy container CPU limit. Added an explanation in the documentation to call out this observation.

Reply:
what were the new container limits you set for the new test?

Reply:
Envoy CPU limit is set to 4 cores.

Reply:
I'm still a bit skeptical of that 4.6ms on the p999 for latency not being artificial due to one of the sidecars getting throttled. Can we try running again, or even further increasing the number of nodes/cpu limits?

At the limits we set, how many pods did we have? did we have enough CPU in the cluster to satisfy all of the max limits for sidecar/controle plane/load gen/and echo application?

Reply:
Tests re-run on the cluster and gets more reasonable results. It seems that cluster performance fluctuate days apart. Document is updated.